gloo/docs/dev/writing_auth_plugins/_index.md

```diff
@@ -0,0 +1,575 @@
+
+```go
+type ExtAuthPlugin interface {
+	NewConfigInstance(ctx context.Context) (configInstance interface{}, err error)
+	GetAuthService(ctx context.Context, configInstance interface{}) (AuthService, error)
+}
+```
+
+Objects that implement this interface are used as factories for authentication service instances. After Gloo detects a 
+reference to your plugin on a Virtual Service and loads it, it will call the `NewConfigInstance` function to get an 
+object to deserialize the plugin configuration into. 
+
+{{% notice warning %}}
+The object returned by the `NewConfigInstance` function **MUST** be a pointer type.
+{{% /notice %}}
+
+Let's see example to understand this better. If your plugin configuration looks like this:
+
+{{< highlight yaml "hl_lines=17-22" >}}
+apiVersion: gateway.solo.io/v1
+kind: VirtualService
+metadata:
+  name: test-auth
+  namespace: gloo-system
+spec:
+  virtualHost:
+    domains: {} # omitted for brevity
+    routes: {} # omitted for brevity
+    virtualHostPlugins:
+      extensions:
+        configs:
+          extauth:
+            plugin_auth:
+              plugins:
+              - name: my-plugin
+                plugin_file_name: MyPlugin.so
+                exported_symbol_name: Plugin
+                config:
+                  some_key: value-1
+                  some_struct:
+                    another_key: value-2
+{{< /highlight >}}
+
+then `NewConfigInstance` function of your `ExtAuthPlugin` implementation should return a pointer to the following Go struct:
+
+```go
+type MyPluginConfig struct {
+    SomeKey string
+    SomeStruct NestedConfig
+}
+
+type NestedConfig struct {
+    AnotherKey string
+ }
+```
+
+Gloo will populate the struct fields with the values found on the correspondent YAML attributes.
+
+{{% notice note %}}
+You might have noticed that the `plugins` attribute in the configuration example above is an array. It is in fact 
+possible to define multiple plugins on a virtual host. We'll see how [plugin chains](#plugin-chains-and-header-propagation) work later on.
+{{% /notice %}}
+
+The `GetAuthService` function will be invoked by Gloo right after this step. As its `configInstance` argument, Gloo will 
+pass the object that it just populated with the values from the plugin configuration. This function must return an 
+instance of the [AuthService](https://github.com/solo-io/ext-auth-plugins/blob/master/api/interface.go#L29) interface.
+
+##### AuthService
+`AuthService` instances are responsible for authorizing individual requests. This is the interface that all of Gloo's 
+out-of-the box auth implementations (basic auth, OIDC, etc.) implement as well. Your plugin is responsible for providing 
+Gloo with a valid `AuthService` implementation.
+
+```go
+type AuthService interface {
+	Start(ctx context.Context) error
+	Authorize(ctx context.Context, request *envoyauthv2.CheckRequest) (*AuthorizationResponse, error)
```

Comment:
just a thought, but wondering if we should expand the interface here to support a shutdown hook as well. not something we need or use currently, but there might be use cases where it may be desirable. but also, YAGNI? ¯\_(ツ)_/¯

open to thoughts here, could go either way

Reply:
I suppose they could also just start a shutdown hook on a goroutine in their `Start` implementation that waits for the context to be cancelled. I'm happy with that

Reply:
The idea is that the plugin developer should use the context to manage shutdown, i.e. they can use the ctx.Done channel as a "shutdown signal". I have included a section on this later in the guide. Is this the use case you had in mind?

Reply:
Just saw the comment :)